packages/gatsby/src/schema/run-query.js

```diff
@@ -0,0 +1,56 @@
+const _ = require(`lodash`)
+const queryLoki = require(`./query-loki`)
+const querySift = require(`./query-sift`)
+const { pluginFieldTracking } = require(`./plugin-fields`)
+
+function hasPluginFields(queryArgs) {
+  return _.some(queryArgs.filter, (v, fieldName) =>
+    pluginFieldTracking.has(fieldName)
+  )
+}
+
+function chooseQueryEngine(queryArgs) {
+  if (hasPluginFields(queryArgs)) {
```

Comment:
This will definitely need to be more robust - our filter schema support is not really that great but even now we can filter by linked node with ___NODE convention (which need to use custom resolver to access underlying node)

Reply:
Great catch. I didn't notice the linked field example. Can you think of any other types of fields that might need to be resolved too? How about mapping fields?

When @KyleAMathews  and I were discussing this PR, we decided that the vast majority of sites would not be using pluginFields. I assume they also aren't using linked fields. But I could be wrong in that assumption

Reply:
We currently don't support filtering on mapped fields (that's what I meant that support isn't great), but it's something that we would definitely would want to be able to do (here's one of the issues where people mention they can't really filter by mapped fields - https://github.com/gatsbyjs/gatsby/issues/7251, I have WIP branch for this - https://github.com/pieh/gatsby/commit/ccd110c0074ec6f7e802447a766b079146432a8a).

For now only ___NODE fields would be problem. Worth looking into date or local File fields which do have custom resolvers for output, but we don't create correct filter schema for those right now. 

When we create schema type fields with custom resolvers we should track those fields (similar to how you track fields from plugins)

Reply:
I hate the "resolve all the things" approach of run-sift, but I get why it exists. It's a shame, because the second we have to resolve all nodes, we lose any ability to perform index based lookups. 

For plugin filter fields, I think this will always be the case since their resolvers can do *anything*. But I wonder if there's an elegant solution for linked and mapped nodes. If we were using a graph database, then we would already have the link setup. 

One way or another, we'll need to find a proper solution eventually, because I imagine many sites will want to take advantage of linked and mapped filtering.